scipy/sparse/compressed.py

```diff
@@ -835,24 +835,37 @@ def _insert_many(self, i, j, x):
 
         self.check_format(full_check=False)
 
-    def _get_single_element(self,row,col):
+    def _get_single_element(self, row, col):
         M, N = self.shape
         if (row < 0):
             row += M
         if (col < 0):
             col += N
         if not (0 <= row < M) or not (0 <= col < N):
-            raise IndexError("index out of bounds")
+            raise IndexError("index out of bounds: 0<=%d<%d, 0<=%d<%d" %
+                             (row, M, col, N))
 
-        major_index, minor_index = self._swap((row,col))
-
-        # TODO make use of sorted indices (if present)
+        major_index, minor_index = self._swap((row, col))
 
         start = self.indptr[major_index]
-        end = self.indptr[major_index+1]
-        # can use np.add(..., where) from numpy 1.7
-        return np.compress(minor_index == self.indices[start:end],
-                           self.data[start:end]).sum(dtype=self.dtype)
+        end = self.indptr[major_index + 1]
+
+        if self.has_sorted_indices:
+            # Copies may be made, if dtypes of indices are not identical
+            minor_index = self.indices.dtype.type(minor_index)
+            minor_indices = self.indices[start:end]
+            insert_pos_left = np.searchsorted(
+                minor_indices, minor_index, side='left')
+            insert_pos_right = insert_pos_left + np.searchsorted(
+                minor_indices[insert_pos_left:], minor_index, side='right')
+            if insert_pos_left == insert_pos_right:
+                return self.dtype.type(0)
+            else:
+                return self.data[start + insert_pos_left]
+        else:
+            # can use np.add(..., where) from numpy 1.7
```

Comment:
SciPy no longer supports versions of NumPy older than 1.7, but I think this comment is misleading... Unless @jnothman wants to chime in, he seems to have had in mind doing something like:

```
np.add.reduce(self.data[start:end], where=(minor_index == self.indices[start:end]), dtype=self.dtype)
```

Unfortunately, the `where=` kwarg is supported by ufuncs, but not by their methods, so the above doesn't work.